pkg/operator/vspherecontroller/vspherecontroller.go

```diff
@@ -0,0 +1,409 @@
+	klog.V(4).Infof("%s: sync started", c.name)
+	defer klog.V(4).Infof("%s: sync complete", c.name)
+	opSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+
+	if opSpec.ManagementState != operatorapi.Managed {
+		klog.V(4).Infof("%s: ManagementState is %s, skipping", c.name, opSpec.ManagementState)
+		return nil
+	}
+
+	infra, err := c.infraLister.Get(infraGlobalName)
+	if err != nil {
+		return err
+	}
+
+	if infra.Status.PlatformStatus == nil {
+		klog.V(4).Infof("Unknown platform: infrastructure status.platformStatus is nil")
+		return nil
+	}
+
+	if infra.Status.PlatformStatus.Type != ocpv1.VSpherePlatformType {
+		klog.V(4).Infof("Unsupported platform: infrastructure status.platformStatus.type is %s", infra.Status.PlatformStatus.Type)
+		return nil
+	}
+
+	connectionResult := c.loginToVcenter(ctx, infra)
+	defer func() {
+		klog.V(4).Infof("%s: vcenter-csi logging out from vcenter", c.name)
+		err := c.vSphereConnection.Logout(ctx)
+		if err != nil {
+			klog.Errorf("%s: error closing connection to vCenter API: %v", c.name, err)
+		}
+	}()
+
+	// if there was an OCP error we should degrade the cluster or if we previously created storageClass
+	// but we can't connect to vcenter now, we should also degrade the cluster
+	err, degradedOrBlockedFromUpgrades := c.blockUpgradeOrDegradeCluster(ctx, connectionResult)
+	if err != nil {
+		return err
+	}
+	if degradedOrBlockedFromUpgrades {
+		return nil
+	}
+
+	// sync storage class
+	// TODO: There are two questions unanswered here:
+	//	1. Should we sync/create storageclass even if cluster check fails?
+	//	2. Should cluster check not be peformed when storage class sync fails?
+	storageClassSyncResult := c.storageClassController.Sync(ctx, c.vSphereConnection)
+
+	err, degradedOrBlockedFromUpgrades = c.blockUpgradeOrDegradeCluster(ctx, storageClassSyncResult)
+	if err != nil {
+		return err
+	}
+	if degradedOrBlockedFromUpgrades {
+		return nil
+	}
+
+	delay, result, checkRan := c.runClusterCheck(ctx)
+
+	// if checks did not run
+	if !checkRan {
+		return nil
+	}
+	queue := syncContext.Queue()
+	queueKey := syncContext.QueueKey()
+
+	klog.V(2).Infof("Scheduled the next check in %s", delay)
+	time.AfterFunc(delay, func() {
+		queue.Add(queueKey)
+	})
+
+	err, degradedOrBlockedFromUpgrades = c.blockUpgradeOrDegradeCluster(ctx, result)
+	if err != nil {
+		return err
+	}
+
+	// if operand was not started previously and block upgrade is false and clusterdegrade is also false
+	// then and only then we should start CSI driver operator
+	if !c.operandControllerStarted && !degradedOrBlockedFromUpgrades {
+		go c.runConditionalController(ctx)
+		c.operandControllerStarted = true
+	}
+	return nil
+}
+
+func (c *VSphereController) blockUpgradeOrDegradeCluster(ctx context.Context, result checks.ClusterCheckResult) (error, bool) {
+	var clusterCondition string
+	blockUpgrade := c.shouldDegradeCluster(result)
```

Comment:
the variable name is misleading

Reply:
this bit of code was rewritten with different logic. PTAL.